swiftnav/src/signal.rs

```diff
@@ -86,6 +87,22 @@ impl Constellation {
         };
         c_str.to_string_lossy()
     }
+
+    /// Attempts to make a [`Constellation`] from a string
+    pub fn from_str<T: Into<Vec<u8>>>(s: T) -> Result<Constellation, InvalidConstellation> {
+        let mut vec = s.into();
+        vec.push(0);
+        let c_str = ffi::CString::from_vec_with_nul(vec).unwrap();
```

Comment:
could do something like 

```rust
    pub fn from_bytes<T: AsRef<[u8]>>(s: T) -> Result<Constellation, InvalidConstellation> {
        let bytes = s.as_ref();
        let bytes = if matches!(bytes.last(), Some(0)) {
            Cow::Borrowed(bytes)
        } else {
            let mut bytes = bytes.to_vec();
            bytes.push(0);
            Cow::Owned(bytes)
        };
        let c_str = ffi::CStr::from_bytes_with_nul(&bytes).unwrap();
        let constellation = unsafe { swiftnav_sys::constellation_string_to_enum(c_str.as_ptr()) };

        Self::from_constellation_t(constellation)
    }
```

Reply:
And we should probably also probably be using little helper functions to reduce the cost of monomorphization

```rust
 /// Attempts to make a [`Constellation`] from bytes
    pub fn from_bytes<T: AsRef<[u8]>>(s: T) -> Result<Constellation, InvalidConstellation> {
        fn inner(bytes: &[u8]) -> Result<Constellation, InvalidConstellation> {
            let bytes = if matches!(bytes.last(), Some(0)) {
                Cow::Borrowed(bytes)
            } else {
                let mut bytes = bytes.to_vec();
                bytes.push(0);
                Cow::Owned(bytes)
            };
            let c_str = ffi::CStr::from_bytes_with_nul(&bytes).unwrap();
            let constellation =
                unsafe { swiftnav_sys::constellation_string_to_enum(c_str.as_ptr()) };
            Constellation::from_constellation_t(constellation)
        }
        inner(s.as_ref())
    }
```

Reply:
Might also make sense to implement FromStr if we have a method to parse from bytes

Reply:
> The intention for these functions is to be an inverse to the `to_str` functions and not to necessarily operate on any arbitrary byte array.

Ahh yeah that makes sense to me

> I somehow missed [CString::new()](https://doc.rust-lang.org/stable/std/ffi/struct.CString.html#method.new) on my first pass which seems to deal with the null terminator problem itself, but also raises a different issue of a null character in the middle of the string.

CString::new() would require taking ownership of the bytes to perform the conversion. If we go with CStr we don't have to move/clone the bytes in the case the string already has a null byte.  CStr doesn't have the equivalent new function but the check I have above should be pretty similar? I guess we shouldn't be unwraping the result from `from_bytes_with_nul` as that might occur because of a null byte in the middle of the string

Reply:
It seems like the [FFI documentation](https://doc.rust-lang.org/stable/std/ffi/index.html#representations-of-non-rust-strings) suggests to use `CString` for converting Rust strings into C compatible strings and `CStr` when converting C strings into something Rust can easily use. I think this is because Rust strings will only very rarely have null bytes at the very end, so you'll always need to copy the string contents over and append a null byte to make is C compatible and `CString::new()` does this automatically for you. The chances of being able to avoid an allocation by using `CStr` seem small.